### DISCUSSION TITLE
How do you think about add "Conventional Commits" to CONTRIBUTING.md?

### DISCUSSION THREAD

**Discussion by tkoyama010** [pyvista/pyvista#1914] [Ideas]
I found [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) as a format of git commit message. How do you think about this? This is just an idea, and I think we should be cautious about forcing the git commit message format on others. However, I like the idea of this project and would like to hear your thoughts on whether to adopt it. I think we can add this to CONTRIBUTING.md

**Comment by akaszynski** [comment]
One of the things that's great about our project is we avoid (almost without exception) commiting directly to the `main` branch. Due to this, our commits have human readable messages with linked web pages (in the form of the pull requests). I think this seems more than sufficient.

Were we to commit directly to main, we would, without a doubt, need to follow a commit message convention. I think we've done a fairly good job about naming our PRs, but perhaps we should define a standard for that?

**Reply by banesullivan** [reply]
I agree 100% here - having informative PR names is all we really need here since each commit links back to a PR where there is a detailed description and conversation. I use this ALL the time to see how/why some change was made and I rarely look at actual commit messages.

**Comment by adeak** [comment]
To add to what @akaszynski said, if raw commit messages _were_ relevant for the project, we wouldn't be able to enforce complex requirements for contributors. At best categorising commits with feat/enh/etc. could get through. We would either have to enforce rules with pre-commit hooks or similar (which would frustrate users until they get it right), or we'd have to ask/force users to force push their feature branches until commit messages are right.

Overall, this sounds like something that's more appropriate for tight-knit groups of experts, rather than an open-source project where we want to keep the barrier of entry as low as possible.

**Comment by adam-grant-hendry** [comment]
Interesting discussion. I agree with all the points: I tried [`commitizen`](https://commitizen-tools.github.io/commitizen/config/) for a while and hit all the pain points mentioned, though it was nice for automated Changelog generation (I'll give [`hallmark cc add`](https://github.com/vweevers/hallmark#usage) a shot now) and auto-bumping versions.

FWIW: https://common-changelog.org/#42-conventional-commits

Thought: What about [adding a `--pattern`](https://www.devwithimagination.com/2020/04/13/git-commit-hooks-for-branch-naming-pre-commit/) to the [`.pre-commit-config.yaml`](https://github.com/pyvista/pyvista/blob/d546beacee5ea2218f333324a9073d61b95a3b04/.pre-commit-config.yaml#L98) to enforce the [branch naming strategy](https://github.com/pyvista/pyvista/blob/main/CONTRIBUTING.rst#branch-naming-conventions)?

**Reply by banesullivan** [reply]
I like this idea of enforcing a branch naming strategy via pre-commit! However this will make creating one-off branches directly on GitHub problematic and in theory our branch naming strategy only applies to this repo, not necessarily forks as most of the benefit is auto-tagging and running specific CI jobs.

**Reply by tkoyama010** [reply]
I'm sorry for suggesting this myself, but I have changed my mind since then. Conventional Commits is for automatic release note creation, but we have already done that with the labels feature and automatic release note generation. I now believe that commit messages should be free and appropriate, as enforcing certain rules on commits would cause extra stress to developers. Of course, the PR title will be used in the release notes, so we should be easy to understand accordingly.

**Reply by adam-grant-hendry** [reply]
No need to be sorry. I agree. Enforcing a type at the commit level is too restrictive. It doesn't make sense. A change involves an entire PR and often these contain multiple different kinds of content changes (features, docs, tests, etc.). The PR encapsulates what the entire change is, which is why it makes sense to suggest branch names in Contributing. I also agree the branch names should be fluid. They shouldn't be strictly enforced, but guidelines only.

It's like @akaszynski and @adeak said: If we didn't have PRs, we would have to have a stricter commit naming/schema strategy and you would probably want to maintain 3-way merge commits to capture change history, which is extremely ugly. You can't `git bisect` through a 3-way merge to find errors when they occur and the nonlinear history with 3-way merges fights against the cleaner CI/CD methodology.

The kind of changes and history of changes, as well as the code review, are all kept in the PRs. Completely agree Conventional Commits happens at too granular a level and hinders development rather than promotes agility.

**Reply by adam-grant-hendry** [reply]
The only thing I would love to have is something that could help quickly write Release Notes, but in the age of ChatGPT that's probably a no-brainer now. Documentation has to be thought through and it takes group work to review, so it's not a great idea to leave it to be wholly automated (though partially might be nice).
